### PR TITLE
MinAreaRect fix rotating calipers

### LINKED ISSUE #19769: minAreaRect angle calculation fails with a particular contour
##### System information (version)
- OpenCV => 4.5.1.48
- Operating System / Platform => Ubuntu 20.04
- Compiler => Python3.8

##### Detailed description

minAreaRect is supposed to provide an angle as part of a rotated bounding box of a contour. It works fine in most cases, but I encountered multiple cases where a wrong angle is returned. This should have been fixed with [!18157](https://github.com/opencv/opencv/issues/18157) but now happens again. Please note that it is important to use the contour provided in the .txt file.

##### Steps to reproduce

Python code:
```
    contour = np.loadtxt("defect_contour.txt").reshape(-1, 1, 2).astype(np.int32)
    rect = cv.minAreaRect(contour) 
    box = cv.boxPoints(rect) 
    box = np.int0(box)    
    out = np.zeros((3000, 3000, 3))
    cv.drawContours(out,[box],-1,(0,255,0),2)
    cv.drawContours(out,[contour],-1,(255,0,0),2)
    cv.imwrite("defect_contour.png", out)
```
Using this contour:
[defect_contour.txt](https://github.com/opencv/opencv/files/6194745/defect_contour.txt)

The resulting image will look like this:
defect_contour

##### Issue submission checklist

 - [X] I report the issue, it's not a question
   
 - [X] I checked the problem with documentation, FAQ, open issues,
       forum.opencv.org, Stack Overflow, etc and have not found solution
   
 - [X] I updated to latest OpenCV version and the issue is still there
   
 - [X] There is reproducer code and related data files: videos, images, onnx, etc

### PR DESCRIPTION
**Pull request by AleksandrPanov**
resolves #19769

To fix for #19769 (also all samples from #11915 work).

1. Error in function `static void rotatingCalipers( const Point2f* points, int n, int mode, float* out )`.
2. Calculating the minimum angle between calipers side and polygon edge using the cosine results in an incorrect angle due to limited arithmetic accuracy.
3. Replaced vector sorting by angle cosine with sorting with custom comparator. The comparator determins vector order by [dot product sign](https://stackoverflow.com/questions/13221873/determining-if-one-2d-vector-is-to-the-right-or-left-of-another/13221874#13221874) to fix floating point accuracy issue with colinear and close to colinear vectors.
4. Added tests "reproducer_19769" and "reproducer_19769_lightweight".

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
allow_multiple_commits=1
```

### CODE REVIEW

**Comment by AleksandrPanov** [comment]
Also created a reproducer_19769_lightweight test from defect_contour.txt (#19769). This test contains only 23 points and reproduces the problem.
convexHull22

The new angle calculation algorithm solves the problem.
convexHull22

Also the algorithm correctly finds the minimum area rectangle.
convexHull9

File: modules/imgproc/src/rotcalipers.cpp [hunk 1]
```diff
@@ -179,32 +206,20 @@ static void rotatingCalipers( const Point2f* points, int n, int mode, float* out
...
-
-        float maxcos = dp[0] * inv_vect_length[seq[0]];
-
         /* number of calipers edges, that has minimal angle with edge */
         int main_element = 0;
 
-        /* choose minimal angle */
-        for ( i = 1; i < 4; ++i )
+        /* choose minimum angle between calipers side and polygon edge */
+        /* calculating the minimum angle using the cosine results in an incorrect angle selection in 18157 test */
```
Comment (asmorkalov) [inline modules/imgproc/src/rotcalipers.cpp:213]: incorrect result is to strong statement. Original solution based on angle comparison produces wrong result due to limited arithmetic accuracy.

Comment (asmorkalov) [reply modules/imgproc/src/rotcalipers.cpp:213]: I propose to replace the comment with more detailed PR description. For example:
"Replaced vector sorting by angle cosine with sorting with custom comparator. The comparator determins vector order by dot project sign to fix floating point accuracy issue with colinear and close to colinear vectors."

Comment (AleksandrPanov) [reply modules/imgproc/src/rotcalipers.cpp]: Fixed the comment in the code. Added your note to PR description (second and third item on the list).

**Review by alalek — APPROVED** [review]
Well done :+1:

**Pull request merged**